symbolic-unwind/src/evaluator.rs

```diff
@@ -0,0 +1,881 @@
+        self.0.fmt(f)
+    }
+}
+
+impl FromStr for Constant {
+    type Err = ExprParsingError<String>;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        parsing::constant_complete(input).map_err(|e| ExprParsingError {
+            kind: e.kind,
+            input: e.input.to_string(),
+        })
+    }
+}
+
+/// A binary operator.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum BinOp {
+    /// Addition.
+    Add,
+    /// Subtraction.
+    Sub,
+    /// Multiplication.
+    Mul,
+    /// Division.
+    Div,
+    /// Remainder.
+    Mod,
+    /// Alignment.
+    ///
+    /// Truncates the first operand to a multiple of the second operand.
+    Align,
+}
+
+impl fmt::Display for BinOp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Add => write!(f, "+"),
+            Self::Sub => write!(f, "-"),
+            Self::Mul => write!(f, "*"),
+            Self::Div => write!(f, "/"),
+            Self::Mod => write!(f, "%"),
+            Self::Align => write!(f, "@"),
+        }
+    }
+}
+
+/// An expression.
+///
+/// This is generic so that different number types can be used.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Expr<T> {
+    /// A base value.
+    Value(T),
+    /// A named constant.
+    Const(Constant),
+    /// A variable.
+    Var(Variable),
+    /// An expression `a b §`, where `§` is a [binary operator](BinOp).
+    Op(Box<Expr<T>>, Box<Expr<T>>, BinOp),
+    /// A dereferenced subexpression.
+    Deref(Box<Expr<T>>),
+}
+
+impl<T: fmt::Display> fmt::Display for Expr<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Value(n) => write!(f, "{}", n),
+            Self::Const(c) => write!(f, "{}", c),
+            Self::Var(v) => write!(f, "{}", v),
+            Self::Op(x, y, op) => write!(f, "{} {} {}", x, y, op),
+            Self::Deref(x) => write!(f, "{} ^", x),
+        }
+    }
+}
+
+impl<T: FromStr> FromStr for Expr<T> {
+    type Err = ExprParsingError<String>;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        parsing::expr_complete(input).map_err(|e| ExprParsingError {
+            kind: e.kind,
+            input: e.input.to_string(),
+        })
+    }
+}
+
+/// An assignment `v e =` where `v` is a [variable](Variable) and `e` is an [expression](Expr).
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Assignment<T>(Variable, Expr<T>);
+
+impl<T: fmt::Display> fmt::Display for Assignment<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{} {} =", self.0, self.1)
+    }
+}
+
+impl<T: FromStr> FromStr for Assignment<T> {
+    type Err = ExprParsingError<String>;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        parsing::assignment_complete(input).map_err(|e| ExprParsingError {
+            kind: e.kind,
+            input: e.input.to_string(),
+        })
+    }
+}
+
+pub mod parsing {
```

Comment:
I'd vote for putting this module in it's own file, this file is large enough

Reply:
Yeah, I think that's a good idea.